numpy/core/src/multiarray/descriptor.c

```diff
@@ -3228,7 +3228,22 @@ arraydescr_richcompare(PyArray_Descr *self, PyObject *other, int cmp_op)
 {
     PyArray_Descr *new = _convert_from_any(other, 0);
     if (new == NULL) {
-        return NULL;
+        /* Cannot convert `other` to dtype */
+
+        PyErr_Clear();
+
+        switch (cmp_op) {
+            case Py_LT:
+            case Py_LE:
+            case Py_EQ:
+            case Py_GT:
+            case Py_GE:
+                Py_RETURN_FALSE;
+            case Py_NE:
+                Py_RETURN_TRUE;
+            default:
+                Py_RETURN_NOTIMPLEMENTED;
```

Comment:
My guess is, we should return `NotImplemented` in all of the cases.

Reply:
Returning `NotImplemented` gives us a `TypeError` again.

Reply:
Oh, do we catch it above somewhere?

Reply:
I think Python returns False and True for `==` and `!=` when both return NotImplemented?

Reply:
Oh sorry my bad. Yeah, it works for `==` and `!=`, and raises `TypeError` for others which makes sense.